File: 904-leaf-similar-trees/leaf-similar-trees.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean leafSimilar(TreeNode root1, TreeNode root2) {

        ArrayList<Integer> one = new ArrayList<>();
        ArrayList<Integer> two = new ArrayList<>();

        dfs(root1,one);
        dfs(root2,two);

        return one.equals(two);
        
    }

    public void dfs(TreeNode node, ArrayList arr){

        if(node == null){
            return;
        }

        if(node.left == null && node.right == null){
            arr.add(node.val);
        }

        if(node.left != null){
            dfs(node.left,arr);
        }

        if(node.right != null){
            dfs(node.right,arr);
        }

    }
}

File: 904-leaf-similar-trees/leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:

        one = []
        two = []

        one = self.dfs(root1,one)
        two = self.dfs(root2,two)

        print(one,two)
        return one == two

    
    def dfs(self,node,arr):

        if node is None:
            return arr
        
        if node.left is None and node.right is None:
            arr.append(node.val)
        
        if node.left:
            self.dfs(node.left,arr)
        
        if node.right:
            self.dfs(node.right,arr)
        
        return arr
        


        